src/runtime/crdt/crdt-collection.ts

```diff
@@ -236,3 +238,72 @@ function dominates(map1: VersionMap, map2: VersionMap): boolean {
   }
   return true;
 }
+
+/**
+ * Converts a simple fast-forward operation into a sequence of regular ops.
+ * Currently only supports converting add ops made by a single actor. Returns
+ * null if it could not simplify the fast-forward operation.
+ */
+export function simplifyFastForwardOp<T>(fastForwardOp: CollectionFastForwardOp<T>): CollectionOperation<T>[] {
+  if (fastForwardOp.removed.length > 0) {
+    // Remove ops can't be replayed in order.
+    return null;
+  }
+  if (fastForwardOp.added.length === 0) {
+    // Just a version bump, no add ops to replay.
+    return null;
+  }
+  const actor = getSingleActorIncrement(fastForwardOp.oldClock, fastForwardOp.newClock);
+  if (actor === null) {
+    return null;
+  }
+  // Sort the add ops in increasing order by the actor's version.
+  const addOps = [...fastForwardOp.added].sort(([elem1, v1], [elem2, v2]) => (v1[actor] || 0) - (v2[actor] || 0));
+  const expectedClock = {...fastForwardOp.oldClock};
```

Comment:
since we know only actor incremented the version, can we just compare its value?
var expectedClock = fastForwardOp.oldClock[actor];
  for (const [elem, version] of addOps) {
    expectedClock++;
    if (expectedClock != version[actor]) {
      ...

all you need to check is that clocks for that actor are sequential, so it boils down to something like this:
const versions = addOps.map(([elem, v]) => v[actor]);
  if (versions!==range(versions[0],versions[0]+versions.length)) {
     return null;
  }

Reply:
Great ideas! I implemented your first suggestion. Except unfortunately I need to return the add ops in sorted order, so I still need to do the weird sort, so I can't do your second suggestion.